scripts/irods/test/test_irepl.py

```diff
@@ -1878,3 +1878,140 @@ def test_irepl_Sf_Rd_foo(self):
+
+        self.target_resource = 'target_resource'
+        self.other_resource = 'other_resource'
+
+        with session.make_session_for_existing_admin() as admin_session:
+            lib.create_ufs_resource(admin_session, self.target_resource, hostname=test.settings.HOSTNAME_2)
+            lib.create_ufs_resource(admin_session, self.other_resource, hostname=test.settings.HOSTNAME_3)
+
+    @classmethod
+    def tearDownClass(self):
+        self.user0.__exit__()
+        self.user1.__exit__()
+
+        with session.make_session_for_existing_admin() as admin_session:
+            admin_session.assert_icommand(['iadmin', 'rmuser', self.user0.username])
+            admin_session.assert_icommand(['iadmin', 'rmuser', self.user1.username])
+            lib.remove_resource(admin_session, self.target_resource)
+            lib.remove_resource(admin_session, self.other_resource)
+
+    def test_permissions_that_do_not_allow_user_to_see_object_results_in_no_replication_and_an_error(self):
+        logical_path = os.path.join(self.user0.session_collection, 'this_object_is_invisible')
+        permissions = [None, 'null', 'read_metadata']
+
+        for permission in permissions:
+            with self.subTest(str(permission)):
+                try:
+                    # Create a data object and make sure it is good.
+                    self.user0.assert_icommand(['itouch', '-R', self.target_resource, logical_path])
+                    self.assertTrue(lib.replica_exists_on_resource(self.user0, logical_path, self.target_resource))
+                    self.assertEqual(str(1), lib.get_replica_status(self.user0, os.path.basename(logical_path), 0))
+
+                    if permission is not None:
+                        self.user0.assert_icommand(['ichmod', permission, self.user1.username, logical_path])
+
+                    # Try to replicate the data object and fail because user1 cannot even see the data object.
+                    self.user1.assert_icommand(
+                        ['irepl', '-R', self.other_resource, logical_path], 'STDERR', 'does not exist')
+                    self.assertTrue(lib.replica_exists_on_resource(self.user0, logical_path, self.target_resource))
+                    self.assertEqual(str(1), lib.get_replica_status(self.user0, os.path.basename(logical_path), 0))
+                    self.assertFalse(lib.replica_exists_on_resource(self.user0, logical_path, self.other_resource))
+
+                finally:
+                    # In the case where this test does not pass (i.e. REGRESSION) it is possible for the replicas to be
+                    # stuck in the intermediate or write-locked status. We set the status at the end here to ensure
+                    # that the object can be removed.
+                    with session.make_session_for_existing_admin() as admin_session:
+                        admin_session.assert_icommand(['ils', '-L', os.path.dirname(logical_path)], 'STDOUT')
```

Comment:
what does this assert?   same for those below that are identical.

Reply:
Just debuggin. I've added a comment for each.